lore/metadata/__init__.py

```diff
@@ -0,0 +1,189 @@
+        credit to Kevin @ StackOverflow
+        from: http://stackoverflow.com/questions/2546207/does-sqlalchemy-have-an-equivalent-of-djangos-get-or-create
+        '''
+        session = Session()
+        instance = session.query(cls).filter_by(**kwargs).first()
+        session.close()
+
+        if not instance:
+            self = cls(**kwargs)
+            self.save()
+        else:
+            self = instance
+
+        return self
+
+    def save(self):
+        session = Session()
+        session.add(self)
+        return session.commit()
+
+    def update(self, **kwargs):
+        for key, value in kwargs.items():
+            self.__dict__[key] = value
+        return self.save()
+
+    def delete(self):
+        session = Session()
+        session.delete(self)
+        return session.commit()
+
+
+class Commit(Crud, Base):
+    sha = Column(String, primary_key=True)
+    created_at = Column(DateTime, nullable=False, default=func.now())
+    updated_at = Column(DateTime, nullable=False, default=func.now())
+    message = Column(String)
+    author_name = Column(String, index=True)
+    author_email = Column(String)
+    fittings = relationship('Fitting', back_populates='commit')
+    snapshots = relationship('Snapshot', back_populates='commit')
+
+    @classmethod
+    def from_git(cls, sha='HEAD'):
+        process = subprocess.Popen([
+            'git',
+            'rev-list',
+            '--format=NAME: %an%nEMAIL: %aE%nDATE: %at%nMESSAGE:%N%B',
+            '--max-count=1',
+            sha,
+        ], stdout=subprocess.PIPE)
+        out, err = process.communicate()
+
+        # If there is no Git repo, exit code will be non-zero
+        if process.returncode == 0:
+            lines = out.strip().decode().split(os.linesep)
+
+            check, sha = lines[0].split('commit ')
+            if check or not sha:
+                logger.error('bad git parse: %s' % out)
+
+            check, author_name = lines[1].split('NAME: ')
+            if check or not author_name:
+                logger.error('bad git parse for NAME: %s' % out)
+
+            check, author_email = lines[2].split('EMAIL: ')
+            if check or not author_email:
+                logger.error('bad git parse for EMAIL: %s' % out)
+
+            check, date = lines[3].split('DATE: ')
+            if check or not date:
+                logger.error('bad git parse for DATE: %s' % out)
+            created_at = datetime.datetime.fromtimestamp(int(date))
+
+            check, message = lines[4], os.linesep.join(lines[5:])
+            if check != 'MESSAGE:' or not message:
+                logger.error('bad git parse for MESSAGE: %s' % out)
+
+            return Commit(
+                sha=sha,
+                author_name=author_name,
+                author_email=author_email,
+                created_at=created_at,
+                message=message
+            )
+        else:
+            return None
+
+
+class Snapshot(Crud, Base):
```

Comment:
@ganesh-krishnan may be I am missing something where are the location of the artifacts in snapshot?

Reply:
@sarchak i have not focused much on the pipeline store. we have a url for the model though and the model contains the pipeline. we should chat about streamlining this better.